niceman/utils.py

```diff
@@ -1136,4 +1136,107 @@ def instantiate_attr_object(item_type, kws):
         # if couldn't figure it out -- just raise original
         raise
 
-lgr.log(5, "Done importing niceman.utils")
\ No newline at end of file
+
+class PathRoot(object):
+    """Find the root of paths based on a predicate function.
+
+    The path -> root mapping is cached across calls.
+
+    Parameters
+    ----------
+    predicate : callable
+        A callable that will be passed a path and should return true
+        if that path should be considered a root.
+    """
+    def __init__(self, predicate):
+        self._pred = predicate
+        self._cache = {}  # path -> root
+
+    def __call__(self, path):
+        """Find root of `path` based on `predicate`.
+
+        Parameters
+        ----------
+        path : str
+            Find this path's root.
+
+        Returns
+        -------
+        str or None
+        """
+        to_cache = []
+        root = None
+        for pth in self._walk_up(path):
+            if pth in self._cache:
+                root = self._cache[pth]
+                break
+
+            to_cache.append(pth)
+
+            if self._pred(pth):
+                root = pth
+                break
+
+        for pth in to_cache:
+            self._cache[pth] = root
```

Comment:
In the call today, @yarikoptic asked why this was a separate loop.
It's because we don't know the value of `root` during the previous
loop.

For example, say we have "/path/to/VENV/subdir2/subdir1/file".  We
first encounter "file", "subdir1", and "subdir2".  Only then do we hit
"VENV", but we want to cache "VENV" as the root of "file", "subdir1",
and "subdir2".

Reply:
thanks, indeed -- I have misread the code a bit there